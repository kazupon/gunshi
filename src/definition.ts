/**
 * The entry for command deifinition.
 *
 * @example
 * ```js
 * import { define } from 'gunshi/definition'
 * ```
 *
 * @module
 */

/**
 * @author kazuya kawaguchi (a.k.a. kazupon)
 * @license MIT
 */

import type { ArgOptions } from 'args-tokens'
import type { Command, CommandLoader, LazyCommand } from './types.ts'

export type { ArgOptions, ArgOptionSchema, ArgValues } from 'args-tokens'

/**
 * Define a {@link Command | command} with type inference
 * @param definition A {@link Command | command} definition
 * @returns A {@link Command | command} definition with type inference
 */
export function define<Options extends ArgOptions = ArgOptions>(
  definition: Command<Options>
): Command<Options> {
  return definition
}

/**
 * Define a {@link LazyCommand | lazy command} with command loader, which is attached with command definition as usage metadata.
 * @param loader A {@link CommandLoader | command loader}
 * @param definition A {@link Command | command} definition
 * @returns A {@link LazyCommand | lazy command} loader
 */
export function lazy<Options extends ArgOptions = ArgOptions>(
  loader: CommandLoader<Options>,
  definition?: Command<Options>
): LazyCommand<Options> {
  if (definition != null) {
    ;(loader as LazyCommand<Options>).commandName = definition.name
    ;(loader as LazyCommand<Options>).description = definition.description
    ;(loader as LazyCommand<Options>).options = definition.options
    ;(loader as LazyCommand<Options>).examples = definition.examples
    ;(loader as LazyCommand<Options>).resource = definition.resource
  }
  return loader
}
