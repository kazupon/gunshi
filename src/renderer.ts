/**
 * The entry point for Gunshi renderer.
 *
 * @example
 * ```js
 * import { renderHeader, renderUsage, renderValidationErrors } from 'gunshi/renderer'
 * ```
 *
 * @module
 */
export { renderHeader } from './renderer/header.ts'
export { renderUsage } from './renderer/usage.ts'
export { renderValidationErrors } from './renderer/validation.ts'
