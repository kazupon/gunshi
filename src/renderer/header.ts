import type { ArgOptions } from 'args-tokens'
import type { CommandContext } from '../types'

/**
 * Render the header
 * @param ctx A {@link CommandContext | command context}
 * @returns A rendered header
 */
export function renderHeader<Options extends ArgOptions = ArgOptions>(
  ctx: Readonly<CommandContext<Options>>
): Promise<string> {
  const title = ctx.env.description || ctx.env.name || ''
  return Promise.resolve(
    title
      ? `${title} (${ctx.env.name || ''}${ctx.env.version ? ` v${ctx.env.version}` : ''})`
      : title
  )
}
