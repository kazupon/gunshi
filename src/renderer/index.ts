export { renderHeader } from './header.ts'
export { renderUsage } from './usage.ts'
export { renderValidationErrors } from './validation.ts'
