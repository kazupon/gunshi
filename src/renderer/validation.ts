import type { ArgOptions } from 'args-tokens'
import type { CommandContext } from '../types.ts'

/**
 * Render the validation errors
 * @param ctx A {@link CommandContext | command context}
 * @param error An {@link AggregateError} of option in `args-token` validation
 * @returns A rendered validation error
 */
export function renderValidationErrors<Options extends ArgOptions = ArgOptions>(
  _ctx: CommandContext<Options>,
  error: AggregateError
): Promise<string> {
  const messages = [] as string[]
  for (const err of error.errors as Error[]) {
    messages.push(err.message)
  }
  return Promise.resolve(messages.join('\n'))
}
