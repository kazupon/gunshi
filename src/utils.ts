import { BUILT_IN_KEY_SEPARATOR, BUILT_IN_PREFIX } from './constants.ts'

import type { ArgOptions } from 'args-tokens'
import type {
  Command,
  Commandable,
  CommandBuiltinOptionsKeys,
  CommandBuiltinResourceKeys,
  GenerateNamespacedKey
} from './types.ts'

export async function resolveLazyCommand<Options extends ArgOptions = ArgOptions>(
  cmd: Commandable<Options>,
  name: string | undefined,
  entry: boolean = false
): Promise<Command<Options>> {
  const resolved = Object.assign(
    create<Command<Options>>(),
    typeof cmd == 'function' ? await cmd() : cmd,
    { default: entry }
  )

  if (resolved.name == null && name) {
    resolved.name = name
  }
  return deepFreeze(resolved)
}

export function resolveBuiltInKey<
  Key extends string = CommandBuiltinOptionsKeys | CommandBuiltinResourceKeys
>(key: Key): GenerateNamespacedKey<Key> {
  return `${BUILT_IN_PREFIX}${BUILT_IN_KEY_SEPARATOR}${key}`
}

export function mapResourceWithBuiltinKey(
  resource: Record<string, string>
): Record<string, string> {
  return Object.entries(resource).reduce((acc, [key, value]) => {
    acc[resolveBuiltInKey(key)] = value
    return acc
  }, create<Record<string, string>>())
}

export function create<T>(obj: object | null = null): T {
  return Object.create(obj) as T
}

export function log(...args: unknown[]): void {
  console.log(...args)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function deepFreeze<T extends Record<string, any>>(obj: T): Readonly<T> {
  if (obj === null || typeof obj !== 'object') {
    return obj
  }

  for (const key of Object.keys(obj)) {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const value = obj[key]
    if (typeof value === 'object' && value !== null) {
      deepFreeze(value)
    }
  }

  return Object.freeze(obj)
}
