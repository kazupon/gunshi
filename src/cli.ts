import { parseArgs, resolveArgs } from 'args-tokens'
import { COMMAND_OPTIONS_DEFAULT, COMMON_OPTIONS } from './constants.ts'
import { createCommandContext } from './context.ts'
import { renderHeader, renderUsage, renderValidationErrors } from './renderer.ts'
import { create, resolveLazyCommand } from './utils.ts'

import type { ArgOptions, ArgToken } from 'args-tokens'
import type { Command, CommandContext, CommandOptions, CommandRunner } from './types.ts'

/**
 * Run the command.
 * @param args Command line arguments
 * @param entry A {@link Command | entry command} or an {@link CommandRunner | inline command runner}
 * @param opts A {@link CommandOptions | command options}
 * @returns A rendered usage or undefined. if you will use {@link CommandOptions.usageSilent} option, it will return rendered usage string.
 */
export async function cli<Options extends ArgOptions = ArgOptions>(
  args: string[],
  entry: Command<Options> | CommandRunner<Options>,
  opts: CommandOptions<Options> = {}
): Promise<string | undefined> {
  const tokens = parseArgs(args)

  const subCommand = getSubCommand(tokens)
  const resolvedCommandOptions = resolveCommandOptions(opts, entry)
  const [name, command] = await resolveCommand(subCommand, entry, resolvedCommandOptions)
  if (!command) {
    throw new Error(`Command not found: ${name || ''}`)
  }

  const options = resolveArgOptions(command.options)

  const { values, positionals, error } = resolveArgs(options, tokens)
  const omitted = !subCommand
  const ctx = await createCommandContext({
    options,
    values,
    positionals,
    args,
    tokens,
    omitted,
    command,
    commandOptions: resolvedCommandOptions
  })

  if (values.version) {
    showVersion(ctx)
    return
  }

  const usageBuffer: string[] = []

  const header = await showHeader(ctx)
  if (header) {
    usageBuffer.push(header)
  }

  if (values.help) {
    const usage = await showUsage(ctx)
    if (usage) {
      usageBuffer.push(usage)
    }
    return usageBuffer.join('\n')
  }

  if (error) {
    await showValidationErrors(ctx, error)
    return
  }

  await command.run(ctx)
}

function resolveArgOptions<Options extends ArgOptions>(options?: Options): Options {
  return Object.assign(create<Options>(), options, COMMON_OPTIONS)
}

function resolveCommandOptions<Options extends ArgOptions>(
  options: CommandOptions<Options>,
  entry: Command<Options> | CommandRunner<Options>
): CommandOptions<Options> {
  const subCommands = new Map(options.subCommands)
  if (typeof entry === 'object' && entry.name) {
    subCommands.set(entry.name, entry)
  }
  const resolvedOptions = Object.assign(
    create<CommandOptions<Options>>(),
    COMMAND_OPTIONS_DEFAULT,
    options,
    {
      subCommands
    }
  ) as CommandOptions<Options>

  return resolvedOptions
}

function getSubCommand(tokens: ArgToken[]): string {
  const firstToken = tokens[0]
  return firstToken &&
    firstToken.kind === 'positional' &&
    firstToken.index === 0 &&
    firstToken.value
    ? firstToken.value
    : ''
}

async function showUsage<Options extends ArgOptions>(
  ctx: CommandContext<Options>
): Promise<string | undefined> {
  if (ctx.env.renderUsage === null) {
    return
  }
  const usage = await (ctx.env.renderUsage || renderUsage)(ctx)
  if (usage) {
    ctx.log(usage)
    return usage
  }
}

function showVersion<Options extends ArgOptions>(ctx: CommandContext<Options>): void {
  ctx.log(ctx.env.version)
}

async function showHeader<Options extends ArgOptions>(
  ctx: CommandContext<Options>
): Promise<string | undefined> {
  if (ctx.env.renderHeader === null) {
    return
  }
  const header = await (ctx.env.renderHeader || renderHeader)(ctx)
  if (header) {
    ctx.log(header)
    ctx.log()
    return header
  }
}

async function showValidationErrors<Options extends ArgOptions>(
  ctx: CommandContext<Options>,
  error: AggregateError
): Promise<void> {
  if (ctx.env.renderValidationErrors === null) {
    return
  }
  const render = ctx.env.renderValidationErrors || renderValidationErrors
  ctx.log(await render(ctx, error))
}

async function resolveCommand<Options extends ArgOptions>(
  sub: string,
  entry: Command<Options> | CommandRunner<Options>,
  options: CommandOptions<Options>
): Promise<[string | undefined, Command<Options> | undefined]> {
  const omitted = !sub
  if (typeof entry === 'function') {
    return [undefined, { run: entry }]
  } else {
    if (omitted) {
      return typeof entry === 'object'
        ? [entry.name, await resolveLazyCommand(entry)]
        : [undefined, undefined]
    } else {
      if (options.subCommands == null) {
        return [sub, undefined]
      }

      const cmd = options.subCommands?.get(sub)

      if (cmd == null) {
        return [sub, undefined]
      }

      return [sub, await resolveLazyCommand(cmd, sub)]
    }
  }
}
