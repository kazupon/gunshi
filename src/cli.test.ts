import { afterEach, describe, expect, test, vi } from 'vitest'
import { defineMockLog } from '../test/utils.ts'
import { cli } from './cli.ts'
import { define } from './definition.ts'
import { renderValidationErrors } from './renderer.ts'

import type { ArgOptions } from 'args-tokens'
import type { Command, CommandOptions } from './types.ts'

afterEach(() => {
  vi.resetAllMocks()
})

describe('execute command', () => {
  test('entry iniline function', async () => {
    const mockFn = vi.fn()
    await cli([], mockFn)
    expect(mockFn).toBeCalled()
  })

  test('entry command', async () => {
    const mockFn = vi.fn()
    await cli([], {
      run: mockFn
    })
    expect(mockFn).toBeCalled()
  })

  test('entry command with name', async () => {
    const mockFn = vi.fn()
    await cli(['dist/'], {
      name: 'publish',
      run: mockFn
    })
    expect(mockFn).toBeCalled()
  })

  test('entry command with options', async () => {
    const mockFn = vi.fn()
    await cli(['--outDir', 'dist/', 'foo', 'bar'], {
      options: {
        outDir: {
          type: 'string',
          short: 'f'
        }
      },
      run: mockFn
    })
    expect(mockFn.mock.calls[0][0].values).toEqual({ outDir: 'dist/' })
    expect(mockFn.mock.calls[0][0].positionals).toEqual(['foo', 'bar'])
  })

  test('entry command without options', async () => {
    const mockFn = vi.fn()
    await cli(['dist/', 'test/'], {
      run: mockFn
    })
    expect(mockFn.mock.calls[0][0].values).toEqual({})
    expect(mockFn.mock.calls[0][0].positionals).toEqual(['dist/', 'test/'])
  })

  test('entry strictly command + sub commands', async () => {
    const mockShow = vi.fn()
    const mockCommand1 = vi.fn()
    const mockCommand2 = vi.fn()
    const show = {
      name: 'show',
      run: mockShow
    }
    const subCommands = new Map()
    subCommands.set('command1', {
      name: 'command1',
      options: {
        foo: {
          type: 'string',
          short: 'f'
        }
      },
      run: mockCommand1
    })
    subCommands.set('command2', {
      name: 'command2',
      options: {
        bar: {
          type: 'number',
          short: 'b'
        }
      },
      run: mockCommand2
    })
    const options = {
      subCommands
    }

    await cli([''], show, options) // omit
    await cli(['show'], show, options)
    await cli(['command1', '--foo', 'foo', 'position1'], show, options)
    await cli(['command2', '--bar=1', 'position2'], show, options)

    expect(mockShow).toBeCalledTimes(2)
    expect(mockCommand1).toBeCalledTimes(1)
    expect(mockCommand1.mock.calls[0][0].values).toEqual({ foo: 'foo' })
    expect(mockCommand1.mock.calls[0][0].positionals).toEqual(['command1', 'position1'])
    expect(mockCommand2).toBeCalledTimes(1)
    expect(mockCommand2.mock.calls[0][0].values).toEqual({ bar: 1 })
    expect(mockCommand2.mock.calls[0][0].positionals).toEqual(['command2', 'position2'])
  })

  test('entry loose command + sub commands', async () => {
    const mockAnonymous = vi.fn()
    const mockShow = vi.fn()
    const mockCommand1 = vi.fn()
    const mockCommand2 = vi.fn()
    // no name command
    const anonymous = {
      run: mockAnonymous
    }
    const subCommands = new Map()
    subCommands.set('show', {
      run: mockShow
    })
    subCommands.set('command1', {
      name: 'command1',
      run: mockCommand1
    })
    subCommands.set('command2', {
      name: 'command2',
      run: mockCommand2
    })
    const options = {
      subCommands
    }

    await cli([''], anonymous, options) // omit
    await cli(['show'], anonymous, options)
    await cli(['command1'], anonymous, options)
    await cli(['command2'], anonymous, options)

    expect(mockAnonymous).toBeCalledTimes(1)
    expect(mockShow).toBeCalledTimes(1)
    expect(mockCommand1).toBeCalledTimes(1)
    expect(mockCommand2).toBeCalledTimes(1)
  })

  test('command not found', async () => {
    const subCommands = new Map()
    subCommands.set('foo', {
      run: vi.fn()
    })
    await expect(async () => {
      await cli(['show'], { run: vi.fn() }, { subCommands })
    }).rejects.toThrowError('Command not found: show')
  })

  test('not registered entry in sub commands', async () => {
    const mockEntry = vi.fn()
    const mockCommand1 = vi.fn()

    const entry = {
      name: 'main',
      run: mockEntry
    }
    const subCommands = new Map()
    subCommands.set('command1', {
      name: 'command1',
      run: mockCommand1
    })
    const options = {
      subCommands
    }

    await cli([''], entry, options)
    await cli(['main'], entry, options)
    await cli(['command1'], entry, options)

    expect(mockEntry).toBeCalledTimes(2)
    expect(mockCommand1).toBeCalledTimes(1)
  })
})

describe('auto generate usage', () => {
  test('inline function', async () => {
    const utils = await import('./utils.ts')
    const log = defineMockLog(utils)
    const renderedUsage = await cli(['-h'], vi.fn())

    const message = log()
    expect(message).toMatchSnapshot()
    expect(message).toMatchSnapshot(renderedUsage)
  })

  test('loosely entry command', async () => {
    const utils = await import('./utils.ts')
    const log = defineMockLog(utils)
    const renderedUsage = await cli(['-h'], {
      options: {
        foo: {
          type: 'string',
          short: 'f'
        }
      },
      run: vi.fn()
    })

    const message = log()
    expect(message).toMatchSnapshot()
    expect(message).toMatchSnapshot(renderedUsage)
  })

  test('strictly entry command', async () => {
    const utils = await import('./utils.ts')
    const log = defineMockLog(utils)
    const renderedUsage = await cli(
      ['-h'],
      {
        options: {
          foo: {
            type: 'string',
            short: 'f',
            description: 'The foo option'
          }
        },
        name: 'command1',
        examples: '# Example 1\n$ gunshi --foo bar\n# Example 2\n$ gunshi -f bar',
        run: vi.fn()
      },
      {
        name: 'gunshi',
        description: 'Modern CLI tool',
        version: '0.0.0',
        usageOptionType: true
      }
    )

    const message = log()
    expect(message).toMatchSnapshot(renderedUsage)
  })

  test('loosely sub commands', async () => {
    const utils = await import('./utils.ts')
    const log = defineMockLog(utils)

    const entryOptions = {
      foo: {
        type: 'string',
        short: 'f'
      }
    } satisfies ArgOptions
    const entry = {
      options: entryOptions,
      run: vi.fn()
    } satisfies Command<typeof entryOptions>

    const command2Options = {
      bar: {
        type: 'number',
        short: 'b',
        default: 42
      }
    } satisfies ArgOptions
    const command2 = {
      options: command2Options,
      run: vi.fn()
    } satisfies Command<typeof command2Options>

    const subCommands = new Map()
    subCommands.set('command2', command2)

    expect(await cli(['-h'], entry, { subCommands })).toMatchSnapshot('main')
    expect(await cli(['command2', '-h'], entry, { subCommands })).toMatchSnapshot('command2')

    const message = log()
    expect(message).toMatchSnapshot()
  })

  test('strictly sub commands', async () => {
    const utils = await import('./utils.ts')
    const log = defineMockLog(utils)

    const entryOptions = {
      foo: {
        type: 'string',
        short: 'f',
        description: 'The foo option'
      }
    } satisfies ArgOptions
    const entry = {
      options: entryOptions,
      name: 'command1',
      examples: '# Example 1\n$ gunshi --foo bar\n# Example 2\n$ gunshi -f bar',
      run: vi.fn()
    } satisfies Command<typeof entryOptions>

    const command2Options = {
      bar: {
        type: 'number',
        short: 'b',
        default: 42,
        description: 'The bar option'
      }
    } satisfies ArgOptions
    const command2 = {
      options: command2Options,
      name: 'command2',
      examples: '# Example 1\n$ gunshi command2 --bar 42\n# Example 2\n$ gunshi command2 -b 42',
      // run: vi.fn()
      run: ctx => {
        console.log(ctx.values)
      }
    } satisfies Command<typeof command2Options>

    const subCommands = new Map()
    subCommands.set('command2', command2)

    const mainUsageRendered = await cli(['-h'], entry, {
      subCommands,
      name: 'gunshi',
      description: 'Modern CLI tool',
      version: '0.0.0',
      leftMargin: 4,
      locale: 'ja-JP',
      middleMargin: 15
    })
    expect(mainUsageRendered).toMatchSnapshot('main')

    const command2UsageRendered = await cli(['command2', '-h'], entry, {
      subCommands,
      name: 'gunshi',
      description: 'Modern CLI tool',
      version: '0.0.0'
    })
    expect(command2UsageRendered).toMatchSnapshot('command2')

    const message = log()
    expect(message).toMatchSnapshot()
  })

  test('locale resource not found', async () => {
    const utils = await import('./utils.ts')
    const log = defineMockLog(utils)
    const mockResource = vi.fn().mockRejectedValue(new Error('Resource not found'))
    await cli(
      ['-h'],
      {
        options: {
          foo: {
            type: 'string',
            short: 'f',
            description: 'The foo option'
          }
        },
        name: 'command1',
        examples: '# Example 1\n$ gunshi --foo bar\n# Example 2\n$ gunshi -f bar',
        resource: mockResource,
        run: vi.fn()
      },
      {
        name: 'gunshi',
        description: 'Modern CLI tool',
        version: '0.0.0',
        locale: 'fr-FR',
        usageOptionType: true
      }
    )

    const message = log()
    expect(message).toMatchSnapshot()
  })
})

describe('custom generate usage', () => {
  test('basic', async () => {
    const utils = await import('./utils.ts')
    const log = defineMockLog(utils)

    const entryOptions = {
      foo: {
        type: 'string',
        short: 'f',
        description: 'this is foo option'
      },
      bar: {
        type: 'boolean',
        required: true,
        description: 'this is bar option'
      },
      baz: {
        type: 'number',
        short: 'b',
        default: 42,
        description: 'this is baz option'
      }
    } satisfies ArgOptions

    const entry = {
      options: entryOptions,
      name: 'command1',
      run: vi.fn()
    } satisfies Command<typeof entryOptions>

    const options = {
      name: 'gunshi',
      description: 'Modern CLI tool',
      version: '0.0.0',
      renderHeader: null, // no header
      renderUsage: ctx => {
        const messages: string[] = []

        // render usage section
        messages.push('Usage:')
        messages.push(`  ${ctx.env.name} [options]`)
        messages.push('')

        // render options section
        messages.push('Options:')
        for (const [key, value] of Object.entries(ctx.options)) {
          const description = value.description || ''
          messages.push(
            `  --${key.padEnd(10)} ${`[${value.type}]`.padEnd(12)}`.padEnd(20) + description
          )
        }
        messages.push('')

        return Promise.resolve(messages.join('\n'))
      },
      renderValidationErrors: async (ctx, error) => {
        // call built-in renderer, and decorate like picocolors
        // return pc.red(await renderValidationErrors(ctx, error))
        const msg = `* ${await renderValidationErrors(ctx, error)} *`
        return ['*'.repeat(msg.length), msg, '*'.repeat(msg.length)].join('\n')
      }
    } satisfies CommandOptions<typeof entryOptions>

    // usage
    await cli(['-h'], entry, options)

    // validation errors
    try {
      await cli([''], entry, options)
    } catch {}

    const message = log()
    expect(message).toMatchSnapshot()
  })
})

test('usageSilent', async () => {
  const utils = await import('./utils.ts')
  const log = defineMockLog(utils)

  const entryOptions = {
    foo: {
      type: 'string',
      short: 'f',
      description: 'this is foo option'
    },
    bar: {
      type: 'boolean',
      required: true,
      description: 'this is bar option'
    },
    baz: {
      type: 'number',
      short: 'b',
      default: 42,
      description: 'this is baz option'
    }
  } satisfies ArgOptions

  const entry = {
    options: entryOptions,
    name: 'command1',
    run: vi.fn()
  } satisfies Command<typeof entryOptions>

  const options = {
    name: 'gunshi',
    description: 'Modern CLI tool',
    version: '0.0.0',
    usageSilent: true
  } satisfies CommandOptions<typeof entryOptions>

  // usage with silent
  const usage = await cli(['-h'], entry, options)
  expect(usage).toMatchSnapshot()

  const stdout = log()
  expect(stdout).toBe('')
})

test('_ (rawArgs)', async () => {
  const args = ['--foo', 'bar', '--baz', 'qux']
  const fn = vi.fn()
  await cli(args, fn)
  expect(fn.mock.calls[0][0]._).toEqual(args)
})

test('tokens', async () => {
  const args = ['--foo', 'bar']
  const fn = vi.fn()
  await cli(args, fn)
  expect(fn.mock.calls[0][0].tokens).toEqual([
    {
      index: 0,
      kind: 'option',
      name: 'foo',
      rawName: '--foo',
      value: undefined,
      inlineValue: undefined
    },
    {
      index: 1,
      kind: 'positional',
      value: 'bar'
    }
  ])
})

test('option grouping', async () => {
  const args = ['-sV']
  const mockFn = vi.fn()
  await cli(args, {
    options: {
      silent: {
        type: 'boolean',
        short: 's'
      },
      verbose: {
        type: 'boolean',
        short: 'V'
      }
    },
    run: mockFn
  })

  expect(mockFn.mock.calls[0][0].values).toEqual({ silent: true, verbose: true })
})

test('rest arguments', async () => {
  const args = ['--foo', 'bar', '--', '--baz', 'qux']
  const mockFn = vi.fn()
  await cli(args, {
    options: {
      foo: {
        type: 'string',
        short: 'f'
      }
    },
    run: mockFn
  })

  expect(mockFn.mock.calls[0][0].rest).toEqual(['--baz', 'qux'])
})

test('negatable options', async () => {
  const args = ['dev', '--bar', '--no-foo']
  await cli(args, {
    options: {
      foo: {
        type: 'boolean',
        negatable: true
      },
      bar: {
        type: 'boolean',
        short: 'b'
      },
      baz: {
        type: 'boolean'
      }
    },
    run: ctx => {
      expect(ctx.positionals).toEqual(['dev'])
      expect(ctx.values).toEqual({ foo: false, bar: true })
    }
  })
})

test('enum options', async () => {
  const utils = await import('./utils.ts')
  const log = defineMockLog(utils)

  // success case
  const options = {
    foo: {
      type: 'enum',
      choices: ['a', 'b', 'c']
    }
  } satisfies ArgOptions
  const mockFn1 = vi.fn()
  await cli(['--foo', 'a'], {
    options,
    run: mockFn1
  })
  expect(mockFn1.mock.calls[0][0].values).toEqual({ foo: 'a' })

  // failure case
  await cli(['--foo', 'z'], {
    options,
    run: vi.fn()
  })
  const stdout = log()
  expect(stdout).toEqual(`Option '--foo' should be chosen from 'enum' ["a", "b", "c"] values`)
})

describe('edge cases', () => {
  test(`'description' option`, async () => {
    const command = define({
      name: 'test',
      description: 'This is a test command',
      options: {
        description: {
          type: 'string',
          short: 'd',
          description: 'This is a description of description option'
        }
      },
      run: vi.fn()
    })

    const utils = await import('./utils.ts')
    const log = defineMockLog(utils)

    await cli(['-h'], command, {
      name: 'gunshi',
      description: 'Modern CLI tool',
      version: '0.0.0'
    })

    const stdout = log()
    expect(stdout).toMatchSnapshot()
  })
})
