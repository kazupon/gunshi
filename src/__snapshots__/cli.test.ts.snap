// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`aute generate usage > inline function 1`] = `
"USAGE:
  COMMAND <OPTIONS> 

OPTIONS:
  -h, --help             Display this help message
  -v, --version          Display this version
"
`;

exports[`aute generate usage > loosely entry command 1`] = `
"USAGE:
  COMMAND <OPTIONS> 

OPTIONS:
  -f, --foo <foo>          
  -h, --help               Display this help message
  -v, --version            Display this version
"
`;

exports[`aute generate usage > loosely sub commands 1`] = `
"USAGE:
  COMMAND <OPTIONS> 

OPTIONS:
  -f, --foo <foo>          
  -h, --help               Display this help message
  -v, --version            Display this version

USAGE:
  COMMAND command2 <OPTIONS>

OPTIONS:
  -b, --bar [bar]          
  -h, --help               Display this help message
  -v, --version            Display this version
"
`;

exports[`aute generate usage > strictly entry command 1`] = `
"Modern CLI tool (gunshi v0.0.0)

USAGE:
  gunshi <OPTIONS> 

OPTIONS:
  -f, --foo <foo>          [string]  The foo option
  -h, --help               [boolean] Display this help message
  -v, --version            [boolean] Display this version

EXAMPLES:
  # Example 1
  $ gunshi --foo bar
  # Example 2
  $ gunshi -f bar
"
`;

exports[`aute generate usage > strictly sub commands 1`] = `
"Modern CLI tool (gunshi v0.0.0)

使い方:
    gunshi [command1] <オプション> 
    gunshi <コマンド>

コマンド:
    command2                 
    command1                 

詳細は、コマンドと\`--help\`フラグを実行してください:
    gunshi command2 --help
    gunshi command1 --help

オプション:
    -f, --foo <foo>               The foo option
    -h, --help                    Display this help message
    -v, --version                 Display this version

例:
    # Example 1
    $ gunshi --foo bar
    # Example 2
    $ gunshi -f bar

Modern CLI tool (gunshi v0.0.0)

USAGE:
  gunshi command2 <OPTIONS>

OPTIONS:
  -b, --bar [bar]          The bar option
  -h, --help               Display this help message
  -v, --version            Display this version

EXAMPLES:
  # Example 1
  $ gunshi command2 --bar 42
  # Example 2
  $ gunshi command2 -b 42
"
`;

exports[`custom generate usage > basic 1`] = `
"Usage:
  gunshi [options]

Options:
  --foo        [string]    this is foo option
  --bar        [boolean]   this is bar option
  --baz        [number]    this is baz option
  --help       [boolean]   Display this help message
  --version    [boolean]   Display this version

*******************************
* Option '--bar'  is required *
*******************************"
`;
